### PR TITLE
lightning: improve region actions for huge region number

### LINKED ISSUE #43510: for large import task, lightning needs to improve region split/scatter/check
## Enhancement

we meet a problem that for a 10 lightning parallel import task, region actions are slow so lightning exits with error.

there are two cases at least

- scattering region timed out and lightning continues to import. And due to region are imbalanced some TiKV nodes run out of disk
- checking region continuous timed out with message "region endKey not equal to next region startKey" and lightning exited

### PR DESCRIPTION
**Pull request by lance6716**
### What problem does this PR solve?

Issue Number: close #43510

Problem Summary:

### What is changed and how it works?

- This PR introduce two config, `tikv-importer.region-split-batch-size` and `tikv-importer.region-split-concurrency`.
   - `tikv-importer.region-split-batch-size` controls the maximum region number in one round of region split+scatter. The default value is 4096
   - `tikv-importer.region-split-concurrency` controls the number of workers that handle the region split/scatter job., The default value is number of CPU cores.
- This PR intrudoce a hidden config `region-check-backoff-limit`. The default value is 1800 and its max backoff interval is 2s, so lightning will wait about 3600s.
- This PR changed some retry strategy. When some of retry jobs made progress, we won't increase the retry counter so won't exit retry due to reach retry counter limit.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: br/pkg/lightning/backend/local/localhelper.go [hunk 1]
```diff
@@ -414,36 +415,48 @@ func (local *Backend) BatchSplitRegions(
...
-
-	return region, newRegions, nil
+		// although it's not PDBatchScanRegion, WaitRegionOnlineBackoffer will only
+		// check this error class so we simply reuse it. Will refine WaitRegionOnlineBackoffer
+		// later
+		failedErr = errors.Annotatef(berrors.ErrPDBatchScanRegion, "scatter region failed")
+		return failedErr
+	}, backoffer)
+
+	// TODO: should we return the error from WithRetry?
```
Comment (D3Hunter) [inline br/pkg/lightning/backend/local/localhelper.go:458]: maybe add a todo "there's still chance we may skip scatter"

File: br/pkg/utils/backoff.go [hunk 2]
```diff
@@ -86,6 +86,11 @@ func (rs *RetryState) RecordRetry() {
 	rs.retryTimes++
 }
 
+// ReduceRetry reduces retry times for 1.
+func (rs *RetryState) ReduceRetry() {
+	rs.retryTimes--
```
Comment (D3Hunter) [inline br/pkg/utils/backoff.go:91]: should we check if `retryTimes > 0`?

**Review by D3Hunter — APPROVED** [review]
rest lgtm

File: br/pkg/utils/backoff.go (hunk 2 above)
Comment (lance6716) [reply br/pkg/utils/backoff.go:91]: retryTimes can be -1 for a short time and soon it will be increased to 0. I think we can tolerate it.

**Comment by lance6716** [comment]
```
[2023/05/17 10:30:55.548 +08:00] [INFO] [duplicate.go:890] ["[detect-dupe] collect duplicate rows from tikv"] [tableName=t1] [tasks=10]
[2023/05/17 10:30:55.550 +08:00] [WARN] [duplicate.go:878] ["[detect-dupe] process remote dupTask encounters error, retrying"] [tableName=t1] [startKey=7480000000000000A95F69800000000000000200] [endKey=7480000000000000A95F698000000000000002FB] [tableID=169] [error="failed to create remote duplicate kv stream: message:\"peer is not leader for region 328, leader may None\" not_leader:<region_id:328 > "] [remainAttempts=5]
...
[2023/05/17 10:30:55.551 +08:00] [WARN] [duplicate.go:878] ["[detect-dupe] process remote dupTask encounters error, retrying"] [tableName=t1] [startKey=7480000000000000AF5F69800000000000000200] [endKey=7480000000000000AF5F698000000000000002FB] [tableID=175] [error="failed to create remote duplicate kv stream: message:\"peer is not leader for region 332, leader may None\" not_leader:<region_id:332 > "] [remainAttempts=1]
[2023/05/17 10:30:55.552 +08:00] [ERROR] [duplicate.go:874] ["[detect-dupe] all attempts to process the remote dupTask have failed"] [tableName=t1] [startKey=7480000000000000A95F69800000000000000200] [endKey=7480000000000000A95F698000000000000002FB] [tableID=169] [error="failed to create remote duplicate kv stream: message:\"peer is not leader for region 328, leader may None\" not_leader:<region_id:328 > "]
[2023/05/17 10:30:55.552 +08:00] [ERROR] [duplicate.go:874] ["[detect-dupe] all attempts to process the remote dupTask have failed"] [tableName=t1] [startKey=7480000000000000AF5F69800000000000000200] [endKey=7480000000000000AF5F698000000000000002FB] [tableID=175] [error="failed to create remote duplicate kv stream: message:\"peer is not leader for region 332, leader may None\" not_leader:<region_id:332 > "]
[2023/05/17 10:30:55.552 +08:00] [ERROR] [duplicate.go:954] ["[detect-dupe] collect remote duplicate keys failed"] [table=t1] [takeTime=3.469236ms] [error="failed to create remote duplicate kv stream: message:\"peer is not leader for region 328, leader may None\" not_leader:<region_id:328 > "]
[2023/05/17 10:30:55.552 +08:00] [ERROR] [backend.go:132] ["[ddl-ingest] start remote duplicate checking"] [error="failed to create remote duplicate kv stream: message:\"peer is not leader for region 328, leader may None\" not_leader:<region_id:328 > "] [errorVerbose="message:\"peer is not leader for region 328, leader may None\" not_leader:<region_id:328 > \ngithub.com/pingcap/errors.AddStack\n\texternal/com_github_pingcap_errors/errors.go:174\ngithub.com/pingcap/errors.Trace\n\texternal/com_github_pingcap_errors/juju_adaptor.go:15\ngithub.com/pingcap/tidb/br/pkg/lightning/backend/local.NewRemoteDupKVStream\n\tbr/pkg/lightning/backend/local/duplicate.go:352\ngithub.com/pingcap/tidb/br/pkg/lightning/backend/local.(*DupeDetector).processRemoteDupTaskOnce.func1.1\n\tbr/pkg/lightning/backend/local/duplicate.go:814\ngithub.com/pingcap/tidb/br/pkg/lightning/backend/local.(*DupeDetector).processRemoteDupTaskOnce.func1\n\tbr/pkg/lightning/backend/local/duplicate.go:827\ngithub.com/pingcap/tidb/br/pkg/utils.(*WorkerPool).Apply.func1\n\tbr/pkg/utils/worker.go:58\nruntime.goexit\n\tsrc/runtime/asm_amd64.s:1598\nfailed to create remote duplicate kv stream"] [table=t1] ["index ID"=2] [stack="github.com/pingcap/tidb/ddl/ingest.(*litBackendCtx).FinishImport\n\tddl/ingest/backend.go:132\ngithub.com/pingcap/tidb/ddl.runIngestReorgJob\n\tddl/index.go:931\ngithub.com/pingcap/tidb/ddl.doReorgWorkForCreateIndex\n\tddl/index.go:855\ngithub.com/pingcap/tidb/ddl.(*worker).onCreateIndex\n\tddl/index.go:684\ngithub.com/pingcap/tidb/ddl.(*worker).runDDLJob\n\tddl/ddl_worker.go:1034\ngithub.com/pingcap/tidb/ddl.(*worker).HandleDDLJobTable\n\tddl/ddl_worker.go:772\ngithub.com/pingcap/tidb/ddl.(*ddl).delivery2worker.func1\n\tddl/job_table.go:422\ngithub.com/pingcap/tidb/util.(*WaitGroupWrapper).Run.func1\n\tutil/wait_group_wrapper.go:154"]
```

**Review by gozssky — APPROVED** [review]

**Comment by lance6716** [comment]
> ```
> [2023/05/17 10:30:55.548 +08:00] [INFO] [duplicate.go:890] ["[detect-dupe] collect duplicate rows from tikv"] [tableName=t1] [tasks=10]
> [2023/05/17 10:30:55.550 +08:00] [WARN] [duplicate.go:878] ["[detect-dupe] process remote dupTask encounters error, retrying"] [tableName=t1] [startKey=7480000000000000A95F69800000000000000200] [endKey=7480000000000000A95F698000000000000002FB] [tableID=169] [error="failed to create remote duplicate kv stream: message:\"peer is not leader for region 328, leader may None\" not_leader:<region_id:328 > "] [remainAttempts=5]
> ...
> [2023/05/17 10:30:55.551 +08:00] [WARN] [duplicate.go:878] ["[detect-dupe] process remote dupTask encounters error, retrying"] [tableName=t1] [startKey=7480000000000000AF5F69800000000000000200] [endKey=7480000000000000AF5F698000000000000002FB] [tableID=175] [error="failed to create remote duplicate kv stream: message:\"peer is not leader for region 332, leader may None\" not_leader:<region_id:332 > "] [remainAttempts=1]
> [2023/05/17 10:30:55.552 +08:00] [ERROR] [duplicate.go:874] ["[detect-dupe] all attempts to process the remote dupTask have failed"] [tableName=t1] [startKey=7480000000000000A95F69800000000000000200] [endKey=7480000000000000A95F698000000000000002FB] [tableID=169] [error="failed to create remote duplicate kv stream: message:\"peer is not leader for region 328, leader may None\" not_leader:<region_id:328 > "]
> [2023/05/17 10:30:55.552 +08:00] [ERROR] [duplicate.go:874] ["[detect-dupe] all attempts to process the remote dupTask have failed"] [tableName=t1] [startKey=7480000000000000AF5F69800000000000000200] [endKey=7480000000000000AF5F698000000000000002FB] [tableID=175] [error="failed to create remote duplicate kv stream: message:\"peer is not leader for region 332, leader may None\" not_leader:<region_id:332 > "]
> [2023/05/17 10:30:55.552 +08:00] [ERROR] [duplicate.go:954] ["[detect-dupe] collect remote duplicate keys failed"] [table=t1] [takeTime=3.469236ms] [error="failed to create remote duplicate kv stream: message:\"peer is not leader for region 328, leader may None\" not_leader:<region_id:328 > "]
> [2023/05/17 10:30:55.552 +08:00] [ERROR] [backend.go:132] ["[ddl-ingest] start remote duplicate checking"] [error="failed to create remote duplicate kv stream: message:\"peer is not leader for region 328, leader may None\" not_leader:<region_id:328 > "] [errorVerbose="message:\"peer is not leader for region 328, leader may None\" not_leader:<region_id:328 > \ngithub.com/pingcap/errors.AddStack\n\texternal/com_github_pingcap_errors/errors.go:174\ngithub.com/pingcap/errors.Trace\n\texternal/com_github_pingcap_errors/juju_adaptor.go:15\ngithub.com/pingcap/tidb/br/pkg/lightning/backend/local.NewRemoteDupKVStream\n\tbr/pkg/lightning/backend/local/duplicate.go:352\ngithub.com/pingcap/tidb/br/pkg/lightning/backend/local.(*DupeDetector).processRemoteDupTaskOnce.func1.1\n\tbr/pkg/lightning/backend/local/duplicate.go:814\ngithub.com/pingcap/tidb/br/pkg/lightning/backend/local.(*DupeDetector).processRemoteDupTaskOnce.func1\n\tbr/pkg/lightning/backend/local/duplicate.go:827\ngithub.com/pingcap/tidb/br/pkg/utils.(*WorkerPool).Apply.func1\n\tbr/pkg/utils/worker.go:58\nruntime.goexit\n\tsrc/runtime/asm_amd64.s:1598\nfailed to create remote duplicate kv stream"] [table=t1] ["index ID"=2] [stack="github.com/pingcap/tidb/ddl/ingest.(*litBackendCtx).FinishImport\n\tddl/ingest/backend.go:132\ngithub.com/pingcap/tidb/ddl.runIngestReorgJob\n\tddl/index.go:931\ngithub.com/pingcap/tidb/ddl.doReorgWorkForCreateIndex\n\tddl/index.go:855\ngithub.com/pingcap/tidb/ddl.(*worker).onCreateIndex\n\tddl/index.go:684\ngithub.com/pingcap/tidb/ddl.(*worker).runDDLJob\n\tddl/ddl_worker.go:1034\ngithub.com/pingcap/tidb/ddl.(*worker).HandleDDLJobTable\n\tddl/ddl_worker.go:772\ngithub.com/pingcap/tidb/ddl.(*ddl).delivery2worker.func1\n\tddl/job_table.go:422\ngithub.com/pingcap/tidb/util.(*WaitGroupWrapper).Run.func1\n\tutil/wait_group_wrapper.go:154"]
> ```

For this problem, I add another commit that increase the retry limit

https://github.com/pingcap/tidb/pull/43807/commits/dec05525cad1bfa5e7c0e934bd904a9160e83216

PTAL @D3Hunter @gozssky

**Review by gozssky — APPROVED** [review]

**Pull request merged**